content/numerical/PolynomialBase.h

```diff
@@ -0,0 +1,57 @@
+/**
+ * Author: chilli, Andrew He, Adamant
+ * Date: 2019-04-27
+ * Description: A FFT based Polynomial class.
+ */
+#pragma once
+
+#include "../number-theory/ModularArithmetic.h"
+#include "FastFourierTransform.h"
+#include "FastFourierTransformMod.h"
+// #include "NumberTheoreticTransform.h"
+
+typedef Mod num;
+typedef vector<num> poly;
+vector<Mod> conv(vector<Mod> a, vector<Mod> b) {
+	auto res = convMod<mod>(vl(all(a)), vl(all(b)));
+	// auto res = conv(vl(all(a)), vl(all(b)));
+	return vector<Mod>(all(res));
+}
+poly &operator+=(poly &a, const poly &b) {
+	a.resize(max(sz(a), sz(b)));
+	rep(i, 0, sz(b)) a[i] = a[i] + b[i];
+	return a;
+}
+poly &operator-=(poly &a, const poly &b) {
+	a.resize(max(sz(a), sz(b)));
+	rep(i, 0, sz(b)) a[i] = a[i] - b[i];
+	return a;
+}
+
+poly &operator*=(poly &a, const poly &b) {
+	if (sz(a) + sz(b) < 100){
+		poly res(sz(a) + sz(b) - 1);
+		rep(i,0,sz(a)) rep(j,0,sz(b))
+			res[i + j] = (res[i + j] + a[i] * b[j]);
+		return (a = res);
+	}
+	return a = conv(a, b);
+}
+poly operator*(poly a, const num b) {
+	poly c = a;
+	trav(i, c) i = i * b;
+	return c;
+}
+#define OP(o, oe) \
+	poly operator o(poly a, poly b) { \
+		poly c = a; \
+		return c oe b; \
+	}
+OP(*, *=) OP(+, +=) OP(-, -=);
```

Comment:
You can get rid of the `oe` parameter, just use `o##=` which concatenates tokens.

```
#define OP(o) \
	poly operator o(poly a, poly b) { \
		poly c = a; \
		return c o##= b; \
	}
OP(+) OP(-) OP(*)
```

Reply:
??? C macros astound me again!

Neat trick.